allennlp/modules/sampled_softmax_loss.py

```diff
@@ -0,0 +1,269 @@
+# https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/python/ops/nn_impl.py#L885
+from typing import Set, Tuple
+
+import numpy as np
+
+import torch
+
+from allennlp.common.checks import ConfigurationError
+
+
+def _choice(num_words: int, num_samples: int) -> Tuple[np.ndarray, int]:
+    """
+    Calls np.random.choice(n_words, n_samples, replace=False),
+    computing the probability on the fly. Returns (samples, num_tries)
+    """
+    num_tries = 0
+    num_chosen = 0
+
+    def get_buffer() -> np.ndarray:
+        log_samples = np.random.rand(num_samples) * np.log(num_words + 1)
+        samples = np.exp(log_samples).astype('int64') - 1
+        return np.clip(samples, a_min=0, a_max=num_words - 1)
+
+    sample_buffer = get_buffer()
+    buffer_index = 0
+    samples: Set[int] = set()
+
+    while num_chosen < num_samples:
+        num_tries += 1
+        # choose sample
+        sample_id = sample_buffer[buffer_index]
+        if sample_id not in samples:
+            samples.add(sample_id)
+            num_chosen += 1
+
+        buffer_index += 1
+        if buffer_index == num_samples:
+            # Reset the buffer
+            sample_buffer = get_buffer()
+            buffer_index = 0
+
+    return np.array(list(samples)), num_tries
+
+
+class SampledSoftmaxLoss(torch.nn.Module):
+    """
+    Based on the default log_uniform_candidate_sampler in tensorflow.
+
+    NOTE num_words DOES NOT include padding id.
+    NOTE: In all cases except (tie_embeddings=True and use_character_inputs=False)
+        the weights are dimensioned as num_words and do not include an entry for the padding (0) id.
```

Comment:
Is this correct in our current setup?

Reply:
I believe so, the second comment sort of doesn't apply, since tie_embeddings isn't implemented, but I don't necessarily want to delete it in case we do implement it later